### ISSUE TITLE
Cookiecutter and the destination folder

### ISSUE THREAD

**Issue by danlester** [Quansight/qhub#573]
I have done some exciting work to move the qhub-terraform-modules into the qhub repo. See the [tfmodules branch](https://github.com/Quansight/qhub/tree/tfmodules)!

It has brought to light an issue around the way cookie-cutter does or does not overwrite destination files.

The immediate 'problem' is that, having added e.g. an `infrastructure/modules` folder to the cookie-cutter template, I find that I don't want it to be Jinja-processed - it will be easier for development, and some of the existing qhub-terraform-modules have code that will be incorrectly interpreted by Jinja. That's fine - they can easily be marked as copy-only in cookie-cutter. Except there is a bug in cookie-cutter that breaks if it does this straight copy to a destination file that already exists.

This is easily solved, so this note isn't really about this problem.

My feeling is that most of the destination folder should be cleared before render.

Here are some example situations:
- In simplifying the GitHub workflows I wanted to delete one of our old workflow files (jupyterhub.yaml since that would actually pick up and `terraform apply` changes already made by the deploy step performed in qhub-ops.yaml). BUT if I just delete it from our template, existing repos will still have the file. (I made a dummy jupyterhub.yaml workflow that does nothing in its place as one solution.)
- We might make changes to the terraform modules (or other templates) which involve deleting files. Which will work for new deployments, but `terraform apply` could pick up old versions of the files on existing deployments.
- Maybe users add random tf files during experimentation and those interfere with our tf structure.

In the tfmodules branch I have a '[cleanout before render](https://github.com/Quansight/qhub/blob/37dfa0f9cf48e0d5c1bfe09290f51ab2f2c43883/qhub/render/__init__.py#L167)' solution that I feel is sensible although complicated if a user needs to really understand what is going on:

- Remove existing folder structure in dest dir apart from:
- Files matching .gitignore entries (e.g. terraform.tfstate) as specified in the TEMPLATE .gitignore file
- Anything the user has added to a .qhubignore file in the dest dir (maybe their own github workflows)
- No FILES in the dest_repo_dir are deleted.
- The .git folder remains intact

The .qhubignore file isn't something that comes with the template, but if the user creates it any files specified won't be deleted. This is useful for e.g. their own GitHub workflows.

My view is that clearing out the existing templates is essential, and I feel this solution is 'natural' if slightly complicated.

Alternatives, assuming we work around the cookie-cutter bug first:
- Keep a closer track of files/folders that we delete over time, so only delete those - i.e. maintain a list of files we have removed from templates. This doesn't solve all problems above.
- For cloud deployments render into a clean tmp folder every time. This would probably mean not committing the terraform files to git. And doesn't solve changes to github workflows.

**Comment by danlester** [comment]
Note that github/gitlab templates are also a consideration, not just terraform though. (But these are much smaller problems.)

**Comment by costrouc** [comment]
My only concern right now with moving towards including the modules inside of the qhub package is that currently one of our clients has a branch off of main to add support for an internal load balancer and fixed ip address. They depend on https://docs.qhub.dev/en/latest/source/02_get_started/04_configuration.html#terraform-modules

**Comment by danlester** [comment]
Of course, if they attempt to upgrade QHub, there could be a mismatch between what qhub is expecting and what their modules are doing. So in that sense they are stuck anyway without readapting to any newer qhub, but if they did that they could certainly want the terraform-modules configuration to still exist.

**Comment by danlester** [comment]
By the way, I think it would be worth it to also respect the terraform_modules setting in qhub-config.yaml if present and do the extra work to pull from the repo for this client.
